Update for Shadowlands

Xruptor opened the pull request:
I added a few missing modules and some code fixes all over the place.  Cleaned up the README and removed from leftover rFilter3 code that wasn't being used.  Added some new examples to the Shaman module of the changes.  Quite a bit of cleanup everywhere really lol.

Xruptor committed:
Added a new glow library for additional glow.
Added the missing CFG to nSpellTracker.lua from the ReadMe.
Fixed the broken Aura scanning.  Now should work properly based on spellID or Name.
Fixed a bug where the Round function was not set to local function and was conflicting with blizzards global Round function.
Added missing Found/NotFound alpha configuration table from the ReadMe.
Added new Glow requirements and features.
Added missing validateUnit option.
Added missing matchSpellID option.
Added missing IsMine option.
Added missing CFG highlightPlayerSpells option from ReadMe.
Added missing CFG refreshInterval option from ReadMe.
Added missing DemonHunter class module file.

Xruptor committed:
ReadMe Fix.

Xruptor committed:
Added ability to track multiple spellID's.  This was apparently removed from the code at some point.
Fixed an issue with GetAlpha not properly taking config alphas into consideration.
Fixed an issue with verifySpell when using multiple spellID's.
Auras will always use the icon texture of the rootSpellID.
Fixed an issue with duration checks.
Added rootSpellID to all functions.  This will act as the primary reference for the trackers in addition to extra spellID's supplied by the configs.
Fixed an issue with the CreateIcon routine.  It was never meant to be used with multiple spellID's.  It must have been modified in the past.
Edited the TOC, it was super out of date and was missing a few fields.
More modifications to the README.  Removing old stuff and putting in new stuff.
Using the Shaman module to test out stuff.
Added initial temporary enchants tracking module.

Xruptor committed:
More updates woot!

Xruptor committed:
Added the ability to track temporary enchants.

Xruptor committed:
Fixed auras routine.
Fixed cooldowns not hiding and changed how they are processed.
Fixed indents all over the place.  I hate spaces!
Changed the not/found alpha to peekAlpha so that it doesn't conflict with regular alpha user setting.
Starting to add spells tracking by usable/notusable or with other checks like mana and what not.

Xruptor committed:
small update to readme

Xruptor committed:
Finalized spell tracking.  WIll add things like mana checks and what not later as I go.
Everything works!

lithammer commented:
> Quite a bit of cleanup everywhere really lol.

No kidding 😮 It's gonna take a bit of time to review. Why is LibCustomGlow necessary?

lithammer commented on libs/LibStub/tests/test4.lua:
```diff
@@ -0,0 +1,41 @@
+assert(LibStub.IterateLibraries == 123)
+
+
+print("OK")
```
Remove the `libs/LibStub/tests` directory.

Xruptor replied:
I'll remove them.

lithammer commented on tempenchants.lua:
```diff
@@ -0,0 +1,138 @@
+	return alpha or 0
+end
+
+local function SetGlow(self, alpha)
```
This function is defined four times.

Xruptor replied:
Consolidated the code.

lithammer commented on cooldowns.lua:
```diff
@@ -1,87 +1,145 @@
+local LibGlow = LibStub("LibCustomGlow-1.0")
+local decimalThreshold = 3
 
 local function GetAlpha(self, duration)
```
This defined four times.

Xruptor replied:
I see that's my bad.  The old copy and paste syndrome got me.

Xruptor replied:
Consolidated the code.

Xruptor commented:
> > Quite a bit of cleanup everywhere really lol.
> 
> No kidding 😮 It's gonna take a bit of time to review. Why is LibCustomGlow necessary? I actually have an updated version locally, just haven't bothered pushing the changes 🤷‍♂️

LibGlow adds the ability to have additional custom action glow animations.  It's also used by WeakAuras.  It's a neat little library.

Xruptor commented:
> > Quite a bit of cleanup everywhere really lol.
> 
> No kidding 😮 It's gonna take a bit of time to review. Why is LibCustomGlow necessary? I actually have an updated version locally, just haven't bothered pushing the changes 🤷‍♂️

Oh in that case, if you would like to use your updated version, I'll retract this pull request before making anymore changes.

lithammer commented:
> > > Quite a bit of cleanup everywhere really lol.
> > 
> > 
> > No kidding 😮 It's gonna take a bit of time to review. Why is LibCustomGlow necessary? I actually have an updated version locally, just haven't bothered pushing the changes 🤷‍♂️
> 
> Oh in that case, if you would like to use your updated version, I'll retract this pull request before making anymore changes. 😄

My changes are only bare minimum (fixing errors) and updated auras for some classes. So I'll definitely take this for a spin when I have some time 🙂 I'm actually surprised anyone was using this (except me of course).

Xruptor commented:
I searched the internet for an addon that was a bare minimum of WeakAuras.  I find the code and options of WeakAuras to be extremely bloated and really only wanted something really simple to track stuff.  I was surprised to find that there are virtually almost no addons out there that do this except yours.  Which I came across.  So I decided to update the addon and provide the missing functionality that I needed.  It worked out great. 😄

Xruptor committed:
A bit of cleanup as requested by the addons original author.
Consolidated some repeating functions.
Removed the test files from LibStub.
Added some more examples to Shaman module.

Xruptor committed:
Added a new setting for the Spell tracker.  You can now set if you want a spell to show if it's on cooldown or not.
The reason behind this is because a spell is treated technically as IsUsable() by the client but it can be on cooldown.  Therefore, if a person only wants a spell to show if it's usable and NOT on cooldown, there was no option for this.

Xruptor commented:
There seems to be a slight bug with the Spells module while tracking if is Usable and on Cooldown.  I believe the global cooldown is causing issues.  I will test it a bit more tomorrow.  I think a combination of IsUsableSpell and GetSpellCooldown is required to circumvent this.  I'll do some trial and error testing later lol.

Xruptor committed:
Fixed issues with the alpha levels sometimes resetting or blinking.
Fixed the Spells module and the global cooldown issues.
Added a new setting (useGlobalCooldown) to allow users to set if they want to use the global cooldown or not for spell cooldowns.
Fixed the issues with (showOnCooldown) and it should now work properly and take spell cooldowns and global cooldowns (if enabled) into consideration.  It will also take spell charges if any into consideration.
Removed from redundant debugging code I accidently left behind.
Updated the Spells module to use events instead of constantly checking it on OnUpdate.  Left the old version just in case for future testing.

Xruptor commented:
Yep it's what I thought.  It was a pesky global cooldown issue.

Xruptor committed:
Alrighty then... the OnUpdate is actually more accurate for Spells module and it isn't as CPU impact as it seems.  I did some testing and it's virtually identical in performance.  The events scanning isn't as reliable as there is a half second delay sometimes.

Xruptor committed:
Tiny enchant code cleanup.

lithammer commented:
Sorry for being slow. Noticed my subscription had ended (haven't logged in for many weeks). Gonna get that sorted.

Xruptor committed:
Adding a bit more fixes for the example shaman class module.

Xruptor committed:
Added a new option for the Spell module to track spells only when they have a Spell Activation Glow Overlay on the actionbars.
Fixed a bug where spells were shown even if they weren't usable.
Fixed a bug where the Spells module was not taking the overrides for the alpha active/inactive settings.
Fixed a bug where enchants were still showing even when the weapons didn't have any.
Added an example to the Death Knight class module to demonstrate the new showOnlyOnOverlayGlow setting.

Xruptor committed:
More alpha fixes.
Fixed a minor issue with the cooldown module and inactivity causing the icon to sometimes show.
Added Hunter example.

Xruptor committed:
forgot showOnCooldown flag.

Xruptor committed:
Don't show Death Knight example out of combat.
